File: ejer2/Java/multiplicacionMatrices.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.Random;

public class multiplicacionMatrices {

    // Función para generar una matriz aleatoria
    public static int[][] generarMatrizAleatoria(int filas, int columnas) {
        int[][] matriz = new int[filas][columnas];
        Random random = new Random();
        for (int i = 0; i < filas; i++) {
            for (int j = 0; j < columnas; j++) {
                matriz[i][j] = random.nextInt(10); // Números aleatorios entre 0 y 9
            }
        }
        return matriz;
    }

    // Función para escribir una matriz en un archivo CSV
    public static void escribirMatrizEnCSV(int[][] matriz, String nombreArchivo, String encabezado) {
        try (FileWriter writer = new FileWriter(nombreArchivo, true)) {
            writer.append(encabezado + "\n");
            for (int[] fila : matriz) {
                for (int i = 0; i < fila.length; i++) {
                    writer.append(String.valueOf(fila[i]));
                    if (i < fila.length - 1) {
                        writer.append(",");
                    }
                }
                writer.append("\n");
            }
            writer.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Función para multiplicar dos matrices y medir el tiempo de ejecución
    public static void multiplicarMatrices(int filasA, int columnasA, int filasB, int columnasB) {
        int[][] matrizA = generarMatrizAleatoria(filasA, columnasA);
        int[][] matrizB = generarMatrizAleatoria(filasB, columnasB);

        // Multiplicar las matrices
        int[][] resultado = new int[filasA][columnasB];
        for (int i = 0; i < filasA; i++) {
            for (int j = 0; j < columnasB; j++) {
                for (int k = 0; k < columnasA; k++) {
                    resultado[i][j] += matrizA[i][k] * matrizB[k][j];
                }
            }
        }

        // Escribir todas las matrices en un archivo CSV
        escribirMatrizEnCSV(matrizA, "resul.csv", "Matriz A (" + filasA + "x" + columnasA + "):");
        escribirMatrizEnCSV(matrizB, "resul.csv", "Matriz B (" + filasB + "x" + columnasB + "):");
        escribirMatrizEnCSV(resultado, "resul.csv", "Resultado de la multiplicacion:");
        
    }

    public static void main(String[] args) {
        int filasA = 200;
        int columnasA = 250;
        int filasB = 250;
        int columnasB = 210;

        // Medir el tiempo de ejecución
        long startTime = System.nanoTime();

        multiplicarMatrices(filasA, columnasA, filasB, columnasB);

        
        // Calcular y mostrar el tiempo de ejecución
        long endTime = System.nanoTime();
        long tiempoNano = endTime - startTime;
        double tiempoMs = tiempoNano / 1_000_000.0; // Convertir nanosegundos a milisegundos
        System.out.println("\nTiempo de ejecución para multiplicar matrices de tamaño "
                + filasA + "x" + columnasA + " y " + filasB + "x" + columnasB + ": " + tiempoMs + " milisegundos");
    }
}

File: ejer2/Ruby/multiplicacionMatrices.rb
require 'csv'
require 'benchmark'

# Función para generar una matriz aleatoriamente
def generar_matriz_aleatoria(filas, columnas)
  Array.new(filas) { Array.new(columnas) { rand(10) } }
end

# Función para escribir varias matrices en un archivo CSV
def escribir_varias_matrices_en_archivo(matrices, nombres, nombre_archivo)
  CSV.open(nombre_archivo, "w") do |csv|
    matrices.each_with_index do |matriz, index|
      csv << [nombres[index]]
      matriz.each { |fila| csv << fila }
      csv << [] # Para dejar una línea en blanco entre cada matriz
    end
  end
end

# Función para multiplicar dos matrices y medir el tiempo de ejecución
def multiplicar_matrices(filas_a, columnas_a, filas_b, columnas_b)
  matriz_a = generar_matriz_aleatoria(filas_a, columnas_a)
  matriz_b = generar_matriz_aleatoria(filas_b, columnas_b)

  resultado = Array.new(filas_a) { Array.new(columnas_b, 0) }

  tiempo = Benchmark.realtime do
    filas_a.times do |i|
      columnas_b.times do |j|
        columnas_a.times do |k|
          resultado[i][j] += matriz_a[i][k] * matriz_b[k][j]
        end
      end
    end
  end

  # Escribir todas las matrices en un solo archivo CSV
  matrices_para_escribir = [matriz_a, matriz_b, resultado]
  nombres_matrices = ["Matriz A (#{filas_a}x#{columnas_a}):", "Matriz B (#{filas_b}x#{columnas_b}):", "Resultado de la multiplicación:"]
  escribir_varias_matrices_en_archivo(matrices_para_escribir, nombres_matrices, 'matrices_result.csv')

  tiempo_ms = (tiempo * 1000).round(4)
  puts "\nTiempo de ejecución para multiplicar matrices de tamaño #{filas_a}x#{columnas_a} y #{filas_b}x#{columnas_b}: #{tiempo_ms} milisegundos"
end

puts "Ingrese el número de filas de la matriz A: "
filas_a = gets.chomp.to_i
puts "Ingrese el número de columnas de la matriz A: "
columnas_a = gets.chomp.to_i
puts "Ingrese el número de filas de la matriz B: "
filas_b = gets.chomp.to_i
puts "Ingrese el número de columnas de la matriz B: "
columnas_b = gets.chomp.to_i

if columnas_a != filas_b
  puts "No se pueden multiplicar las matrices. Las columnas de A deben ser igual a las filas de B."
  exit
end

multiplicar_matrices(filas_a, columnas_a, filas_b, columnas_b)
